ambry-frontend/src/main/java/com/github/ambry/frontend/DeleteBlobHandler.java

```diff
@@ -0,0 +1,167 @@
+        Callback<Void> finalCallback) {
+      this.restRequest = restRequest;
+      this.restResponseChannel = restResponseChannel;
+      this.finalCallback = finalCallback;
+    }
+
+    /**
+     * Start the chain by calling {@link SecurityService#processRequest}.
+     */
+    private void start() {
+      restRequest.setArg(RestUtils.InternalKeys.KEEP_ALIVE_ON_ERROR_HINT, true);
+      securityService.processRequest(restRequest, securityProcessRequestCallback());
+    }
+
+    /**
+     * After {@link SecurityService#processRequest} finishes, call {@link IdConverter#convert} to convert the incoming
+     * ID if required.
+     * @return a {@link Callback} to be used with {@link SecurityService#processRequest}.
+     */
+    private Callback<Void> securityProcessRequestCallback() {
+      return buildCallback(metrics.deleteBlobSecurityProcessRequestMetrics, result -> {
+        String blobIdStr = getRequestPath(restRequest).getOperationOrBlobId(false);
+        idConverter.convert(restRequest, blobIdStr, idConverterCallback());
+      }, restRequest.getUri(), LOGGER, finalCallback);
+    }
+
+    /**
+     * After {@link IdConverter#convert} finishes, call {@link SecurityService#postProcessRequest} to perform
+     * request time security checks that rely on the request being fully parsed and any additional arguments set.
+     * @return a {@link Callback} to be used with {@link IdConverter#convert}.
+     */
+    private Callback<String> idConverterCallback() {
+      return buildCallback(metrics.deleteBlobIdConversionMetrics, convertedBlobId -> {
+        BlobId blobId = FrontendUtils.getBlobIdFromString(convertedBlobId, clusterMap);
+        if (restRequest.getArgs().get(InternalKeys.TARGET_ACCOUNT_KEY) == null) {
+          // Inject account and container when they are missing from the rest request.
+          accountAndContainerInjector.injectTargetAccountAndContainerFromBlobId(blobId, restRequest,
+              metrics.deleteBlobMetricsGroup);
+        }
+        securityService.postProcessRequest(restRequest, securityPostProcessRequestCallback(blobId));
+      }, restRequest.getUri(), LOGGER, finalCallback);
+    }
+
+    /**
+     * After {@link SecurityService#postProcessRequest} finishes, call {@link Router#deleteBlob} to delete
+     * the blob in the storage layer.
+     * @param blobId the {@link BlobId} to undelete
+     * @return a {@link Callback} to be used with {@link SecurityService#postProcessRequest}.
+     */
+    private Callback<Void> securityPostProcessRequestCallback(BlobId blobId) {
+      return buildCallback(metrics.deleteBlobSecurityPostProcessRequestMetrics, result -> {
+        String serviceId = RestUtils.getHeader(restRequest.getArgs(), RestUtils.Headers.SERVICE_ID, false);
+        router.deleteBlob(blobId.getID(), serviceId, routerCallback(),
+            QuotaUtils.buildQuotaChargeCallback(restRequest, quotaManager, false));
+      }, restRequest.getUri(), LOGGER, finalCallback);
```

Comment:
Looks like current code [calculates](https://github.com/linkedin/ambry/blob/master/ambry-frontend/src/main/java/com/github/ambry/frontend/FrontendRestRequestService.java#L1006) charges incurred for DELETE operations on router exceptions too. Would we need same here?

Reply:
above.